README.md
Kodi LanetTV Add-on
===================

This is a PVR add-on for Kodi. It add support for Lanet TV watching.

------------------------------------------

Written by: Anton Baranov

Based on: XBMC PVR Demo Add-on by Pulse-Eight (http://www.pulse-eight.com/)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
See the file COPYING for more information.

---------------------------------------------




changelog.txt
v0.0.4
- rec.lanet.tv removed, not supported in lanet.tv api.
- change api url, because lanet.tv update platform.
- removed groups, old scheme is not supported in lanet.tv api
- update windows build scripts
- update addon.xml
- change git url

v0.0.3 rc1
- Full refactoring code.
- Add records for 2 last days from http://rec.lanet.tv.
- move to Kodi platform (dont work on XBMC)

v0.0.2
- change XML to JSON.
- Refactoring code.
- Rewritten loading STV.
- Removed unused functions.
- Fixed some cases crash.

v0.0.1
- devel start.


src/lanettv/client.h
#pragma once
/*
*      Copyright (C) 2014-2015 Anton Baranov
*      http://cryol.kiev.ua/
*
*      Copyright (C) 2011 Pulse-Eight
*      http://www.pulse-eight.com/
*
*  This Program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2, or (at your option)
*  any later version.
*
*  This Program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with Kodi; see the file COPYING.  If not, write to
*  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
*  http://www.gnu.org/copyleft/gpl.html
*
*/

#include "libXBMC_addon.h"
#include "libXBMC_pvr.h"

#define PVR_CLIENT_VERSION	"0.0.4"
#define GIT_PLUGIN_URL    	"http://kodi.lanet.tv/"

extern bool							m_bCreated;
extern std::string					ApiUrl;
extern std::string					StaticUrl;
extern std::string					g_strEPGPath;
extern std::string					g_strClientPath;
extern std::string					g_strListPath;
extern std::string					g_strLogoPath;
extern ADDON::CHelper_libXBMC_addon *XBMC;
extern CHelper_libXBMC_pvr          *PVR;

src/lanettv/client.cpp
/*
*      Copyright (C) 2014-2015 Anton Baranov
*      http://cryol.kiev.ua/
*
*      Copyright (C) 2011 Pulse-Eight
*      http://www.pulse-eight.com/
*
*  This Program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2, or (at your option)
*  any later version.
*
*  This Program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with Kodi; see the file COPYING.  If not, write to
*  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
*  http://www.gnu.org/copyleft/gpl.html
*
*/

#include "client.h"
#include "xbmc_pvr_dll.h"
#include "libKODI_guilib.h"
#include "PVRLanetTVData.h"
#include "platform/util/util.h"

using namespace std;
using namespace ADDON;

#ifdef TARGET_WINDOWS
#define snprintf _snprintf
#endif

bool				m_bCreated  = false;
ADDON_STATUS		m_CurStatus = ADDON_STATUS_UNKNOWN;
PVRLanetTVData		*m_data = NULL;
bool				m_bIsPlaying = false;
PVRLanetTVChannel	m_currentChannel;

/* User adjustable settings are saved here.
 * Default values are defined inside client.h
 * and exported to the other source files.
 */

std::string g_strUserPath = "";
std::string g_strClientPath	= "";

CHelper_libXBMC_addon *XBMC  = NULL;
CHelper_libXBMC_pvr   *PVR = NULL;

std::string		ApiUrl = "https://api.lanet.tv/";
std::string		StaticUrl = "https://static.lanet.ua/";
std::string		g_strLogoPath = StaticUrl + "tv/logo/";
std::string		g_strEPGPath = ApiUrl + "epg/";
std::string		g_strListPath = ApiUrl + "list.json";

int				g_iEPGTimeShift = 0;
int				g_iStartNumber = 1;
bool			g_bTSOverride = true;

extern "C" {

void ADDON_ReadSettings(void)
{
	XBMC->Log(LOG_DEBUG, "%s - Load addon settings", __FUNCTION__);
  //STUB
}

ADDON_STATUS ADDON_Create(void* hdl, void* props)
{
  if (!hdl || !props)
    return ADDON_STATUS_UNKNOWN;

  PVR_PROPERTIES* pvrprops = (PVR_PROPERTIES*)props;

  XBMC = new CHelper_libXBMC_addon;
  if (!XBMC->RegisterMe(hdl))
  {
    SAFE_DELETE(XBMC);
    return ADDON_STATUS_PERMANENT_FAILURE;
  }

  PVR = new CHelper_libXBMC_pvr;
  if (!PVR->RegisterMe(hdl))
  {
    SAFE_DELETE(PVR);
    SAFE_DELETE(XBMC);
    return ADDON_STATUS_PERMANENT_FAILURE;
  }

  XBMC->Log(LOG_DEBUG, "%s - Creating the PVR LanetTV add-on", __FUNCTION__);

  m_CurStatus     = ADDON_STATUS_UNKNOWN;
  g_strUserPath   = pvrprops->strUserPath;
  g_strClientPath = pvrprops->strClientPath;

  ADDON_ReadSettings();

  m_data = new PVRLanetTVData;
  m_CurStatus = ADDON_STATUS_OK;
  m_bCreated = true;
  return m_CurStatus;
}

ADDON_STATUS ADDON_GetStatus()
{
  return m_CurStatus;
}

void ADDON_Destroy()
{
  delete m_data;
  m_bCreated = false;
  m_CurStatus = ADDON_STATUS_UNKNOWN;
}

bool ADDON_HasSettings()
{
  return true;
}

unsigned int ADDON_GetSettings(ADDON_StructSetting ***sSet)
{
  return 0;
}

ADDON_STATUS ADDON_SetSetting(const char *settingName, const void *settingValue)
{
  return ADDON_STATUS_OK;
}

void ADDON_Stop()
{
}

void ADDON_FreeSettings()
{
}

void ADDON_Announce(const char *flag, const char *sender, const char *message, const void *data)
{
}

/***********************************************************
 * PVR Client AddOn specific public library functions
 ***********************************************************/

const char* GetPVRAPIVersion(void)
{
  static const char *strApiVersion = XBMC_PVR_API_VERSION;
  return strApiVersion;
}

const char* GetMininumPVRAPIVersion(void)
{
  static const char *strMinApiVersion = XBMC_PVR_MIN_API_VERSION;
  return strMinApiVersion;
}

const char* GetGUIAPIVersion(void)
{
  static const char *strGuiApiVersion = KODI_GUILIB_API_VERSION;
  return strGuiApiVersion;
}

const char* GetMininumGUIAPIVersion(void)
{
  static const char *strMinGuiApiVersion = KODI_GUILIB_MIN_API_VERSION;
  return strMinGuiApiVersion;
}

PVR_ERROR GetAddonCapabilities(PVR_ADDON_CAPABILITIES* pCapabilities)
{
  pCapabilities->bSupportsEPG				= true;
  pCapabilities->bSupportsTV				= true;
  pCapabilities->bSupportsRadio				= false;
  pCapabilities->bSupportsChannelGroups		= false;
  pCapabilities->bSupportsRecordings		= false;
  pCapabilities->bSupportsTimers			= false;

  return PVR_ERROR_NO_ERROR;
}

const char *GetBackendName(void)
{
  static const char *strBackendName = "LanetTV PVR Add-on";
  return strBackendName;
}

const char *GetBackendVersion(void)
{
	static CStdString strBackendVersion = PVR_CLIENT_VERSION;
	return strBackendVersion.c_str();
}

const char *GetConnectionString(void)
{
  static CStdString strConnectionString = "connected";
  return strConnectionString.c_str();
}

const char *GetBackendHostname(void)
{
	return "";
}

PVR_ERROR GetDriveSpace(long long *iTotal, long long *iUsed)
{
  *iTotal = 0;
  *iUsed  = 0;
  return PVR_ERROR_NO_ERROR;
}

PVR_ERROR GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd)
{
	if (m_data)
		return m_data->GetEPGForChannel(handle, channel, iStart, iEnd);

	return PVR_ERROR_SERVER_ERROR;
}

int GetChannelsAmount(void)
{
	if (m_data)
		return m_data->GetChannelsAmount();
	return -1;
}

PVR_ERROR GetChannels(ADDON_HANDLE handle, bool bRadio)
{
	if (m_data)
		return m_data->GetChannels(handle, bRadio);

	return PVR_ERROR_SERVER_ERROR;
}

bool OpenLiveStream(const PVR_CHANNEL &channel)
{
  if (m_data)
  {
    CloseLiveStream();

    if (m_data->GetChannel(channel, m_currentChannel))
    {
      m_bIsPlaying = true;
      return true;
    }
  }

  return false;
}

void CloseLiveStream(void)
{
  m_bIsPlaying = false;
}

int GetCurrentClientChannel(void)
{
  return m_currentChannel.iUniqueId;
}

bool SwitchChannel(const PVR_CHANNEL &channel)
{
  CloseLiveStream();

  return OpenLiveStream(channel);
}

PVR_ERROR GetStreamProperties(PVR_STREAM_PROPERTIES* pProperties)
{
  return PVR_ERROR_NOT_IMPLEMENTED;
}


PVR_ERROR SignalStatus(PVR_SIGNAL_STATUS &signalStatus)
{
  snprintf(signalStatus.strAdapterName, sizeof(signalStatus.strAdapterName), "LanetTV Adapter 1");
  snprintf(signalStatus.strAdapterStatus, sizeof(signalStatus.strAdapterStatus), "OK");

  return PVR_ERROR_NO_ERROR;
}


/** UNUSED API FUNCTIONS */
int GetTimersAmount(void) { return -1; }
PVR_ERROR GetRecordings(ADDON_HANDLE handle, bool deleted) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR GetChannelGroups(ADDON_HANDLE handle, bool bRadio) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROUP &group) { return PVR_ERROR_SERVER_ERROR;  }
int GetRecordingsAmount(bool deleted) { return -1; }
int GetChannelGroupsAmount(void) { return -1; }
PVR_ERROR GetTimers(ADDON_HANDLE handle) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR OpenDialogChannelScan(void) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR CallMenuHook(const PVR_MENUHOOK &menuhook, const PVR_MENUHOOK_DATA &item) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR DeleteChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR RenameChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR MoveChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR OpenDialogChannelSettings(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR OpenDialogChannelAdd(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
bool OpenRecordedStream(const PVR_RECORDING &recording) { return false; }
void CloseRecordedStream(void) {}
int ReadRecordedStream(unsigned char *pBuffer, unsigned int iBufferSize) { return 0; }
long long SeekRecordedStream(long long iPosition, int iWhence /* = SEEK_SET */) { return 0; }
long long PositionRecordedStream(void) { return -1; }
long long LengthRecordedStream(void) { return 0; }
void DemuxReset(void) {}
void DemuxFlush(void) {}
int ReadLiveStream(unsigned char *pBuffer, unsigned int iBufferSize) { return 0; }
long long SeekLiveStream(long long iPosition, int iWhence /* = SEEK_SET */) { return -1; }
long long PositionLiveStream(void) { return -1; }
long long LengthLiveStream(void) { return -1; }
const char * GetLiveStreamURL(const PVR_CHANNEL &channel) { return ""; }
PVR_ERROR DeleteRecording(const PVR_RECORDING &recording) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR RenameRecording(const PVR_RECORDING &recording) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR SetRecordingPlayCount(const PVR_RECORDING &recording, int count) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR SetRecordingLastPlayedPosition(const PVR_RECORDING &recording, int lastplayedposition) { return PVR_ERROR_NOT_IMPLEMENTED; }
int GetRecordingLastPlayedPosition(const PVR_RECORDING &recording) { return -1; }
PVR_ERROR GetRecordingEdl(const PVR_RECORDING&, PVR_EDL_ENTRY[], int*) { return PVR_ERROR_NOT_IMPLEMENTED; };
PVR_ERROR AddTimer(const PVR_TIMER &timer) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR DeleteTimer(const PVR_TIMER &timer, bool bForceDelete) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR UpdateTimer(const PVR_TIMER &timer) { return PVR_ERROR_NOT_IMPLEMENTED; }
void DemuxAbort(void) {}
DemuxPacket* DemuxRead(void) { return NULL; }
unsigned int GetChannelSwitchDelay(void) { return 0; }
void PauseStream(bool bPaused) {}
bool CanPauseStream(void) { return false; }
bool CanSeekStream(void) { return false; }
bool SeekTime(int,bool,double*) { return false; }
void SetSpeed(int) {};
time_t GetPlayingTime() { return 0; }
time_t GetBufferTimeStart() { return 0; }
time_t GetBufferTimeEnd() { return 0; }
PVR_ERROR UndeleteRecording(const PVR_RECORDING& recording) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR DeleteAllRecordingsFromTrash() { return PVR_ERROR_NOT_IMPLEMENTED; }
bool IsTimeshifting(void) { return false; }
}


src/lanettv/PVRLanetTVData.h
#pragma once
/*
*      Copyright (C) 2014-2015 Anton Baranov
*      http://cryol.kiev.ua/
*
*      Copyright (C) 2011 Pulse-Eight
*      http://www.pulse-eight.com/
*
*  This Program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2, or (at your option)
*  any later version.
*
*  This Program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with Kodi; see the file COPYING.  If not, write to
*  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
*  http://www.gnu.org/copyleft/gpl.html
*
*/

#include <vector>
#include "platform/util/StdString.h"
#include "client.h"

struct PVRLanetTVEpgEntry
{
	int				iBroadcastId;
	std::string		strTitle;
	int				iChannelId;
	time_t			startTime;
	time_t			endTime;
	std::string		strPlotOutline;
	std::string		strPlot;
	std::string		strIconPath;
	int				iGenreType;
	int				iGenreSubType;
	std::string		strGenreDescription;
	int				iStarRating;
};

struct PVRLanetTVChannel
{
	int								iUniqueId;
	int								iChannelNumber;
	int								iSubChannelNumber;
	int								iEncryptionSystem;
	std::string						strChannelName;
	std::string						strChannelEPG;
	std::string						strIconPath;
	std::string						strStreamURL;
};


class PVRLanetTVData
{
	public:
		PVRLanetTVData(void);
		virtual ~PVRLanetTVData(void);

		virtual PVR_ERROR GetChannels(ADDON_HANDLE handle, bool bRadio);
		virtual bool GetChannel(const PVR_CHANNEL &channel, PVRLanetTVChannel &myChannel);
		virtual int GetChannelsAmount(void);
	
		virtual PVR_ERROR GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd);

	protected:
		virtual bool LoadChannelsData(void);

		virtual std::string LoadFile(std::string url);

	private:
		std::vector<PVRLanetTVChannel>	m_channels;
		time_t							m_iEpgStart;
		CStdString						m_strDefaultIcon;
		CStdString						m_strDefaultMovie;
		std::string						m_strListPack;
};


repository.lanet.ua/addons_xml_generator.py
# *
# *  Copyright (C) 2012-2013 Garrett Brown
# *  Copyright (C) 2010      j48antialias
# *
# *  This Program is free software; you can redistribute it and/or modify
# *  it under the terms of the GNU General Public License as published by
# *  the Free Software Foundation; either version 2, or (at your option)
# *  any later version.
# *
# *  This Program is distributed in the hope that it will be useful,
# *  but WITHOUT ANY WARRANTY; without even the implied warranty of
# *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# *  GNU General Public License for more details.
# *
# *  You should have received a copy of the GNU General Public License
# *  along with XBMC; see the file COPYING.  If not, write to
# *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
# *  http://www.gnu.org/copyleft/gpl.html
# *
# *  Based on code by j48antialias:
# *  https://anarchintosh-projects.googlecode.com/files/addons_xml_generator.py

""" addons.xml generator """

import os
import sys

# Compatibility with 3.0, 3.1 and 3.2 not supporting u"" literals
if sys.version < '3':
    import codecs
    def u(x):
        return codecs.unicode_escape_decode(x)[0]
else:
    def u(x):
        return x

class Generator:
    """
        Generates a new addons.xml file from each addons addon.xml file
        and a new addons.xml.md5 hash file. Must be run from the root of
        the checked-out repo. Only handles single depth folder structure.
    """
    def __init__( self ):
        # generate files
        self._generate_addons_file()
        self._generate_md5_file()
        # notify user
        print("Finished updating addons xml and md5 files")

    def _generate_addons_file( self ):
        # addon list
        addons = os.listdir( "." )
        # final addons text
        addons_xml = u("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<addons>\n\t")
        # loop thru and add each addons addon.xml file
        for addon in addons:
            try:
                # skip any file or .svn folder or .git folder
                if ( not os.path.isdir( addon ) or addon == ".svn" or addon == ".git" ): continue
                # create path
                _path = os.path.join( addon, "addon.xml" )
                # split lines for stripping
                xml_lines = open( _path, "r" ).read().splitlines()
                # new addon
                addon_xml = ""
                # loop thru cleaning each line
                for line in xml_lines:
                    # skip encoding format line
                    if ( line.find( "<?xml" ) >= 0 ): continue
                    # add line
                    if sys.version < '3':
                        addon_xml += unicode( line.rstrip() + "\n\t", "UTF-8" )
                    else:
                        addon_xml += line.rstrip() + "\n\t"
                # we succeeded so add to our final addons.xml text
                addons_xml += addon_xml.rstrip() + "\n\n\t"
            except Exception as e:
                # missing or poorly formatted addon.xml
                print("Excluding %s for %s" % ( _path, e ))
        # clean and add closing tag
        addons_xml = addons_xml.strip() + u("\t\n</addons>\n")
        # save file
        self._save_file( addons_xml.encode( "UTF-8" ), file="addons.xml" )

    def _generate_md5_file( self ):
        # create a new md5 hash
        try:
            import md5
            m = md5.new( open( "addons.xml", "r" ).read() ).hexdigest()
        except ImportError:
            import hashlib
            m = hashlib.md5( open( "addons.xml", "r", encoding="UTF-8" ).read().encode( "UTF-8" ) ).hexdigest()

        # save file
        try:
            self._save_file( m.encode( "UTF-8" ), file="addons.xml.md5" )
        except Exception as e:
            # oops
            print("An error occurred creating addons.xml.md5 file!\n%s" % e)

    def _save_file( self, data, file ):
        try:
            # write data to the file (use b for Python 3)
            open( file, "wb" ).write( data )
        except Exception as e:
            # oops
            print("An error occurred saving %s file!\n%s" % ( file, e ))


if ( __name__ == "__main__" ):
    # start
    Generator()


src/lanettv/PVRLanetTVData.cpp
/*
*      Copyright (C) 2014-2015 Anton Baranov
*      http://cryol.kiev.ua/
*
*      Copyright (C) 2011 Pulse-Eight
*      http://www.pulse-eight.com/
*
*  This Program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2, or (at your option)
*  any later version.
*
*  This Program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with Kodi; see the file COPYING.  If not, write to
*  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
*  http://www.gnu.org/copyleft/gpl.html
*
*/

#include "PVRLanetTVData.h"
#include <json/json.h>
#include <sstream>

using namespace std;
using namespace ADDON;

PVRLanetTVData::PVRLanetTVData(void)
{
	m_iEpgStart = -1;
	m_strDefaultIcon = std::string(GIT_PLUGIN_URL) + "/theme/images/icon.png";
	m_strDefaultMovie = "";

	XBMC->Log(LOG_DEBUG, "%s - Start load data", __FUNCTION__);

	if (LoadChannelsData())
	{
		XBMC->QueueNotification(QUEUE_INFO, "%d channels loaded.", m_channels.size());
	}
}

PVRLanetTVData::~PVRLanetTVData(void)
{
	m_channels.clear();
}

std::string PVRLanetTVData::LoadFile(std::string url)
{
	XBMC->Log(LOG_DEBUG, "Load File: %s", url.c_str());
	std::string retData;

	void* fileHandle = XBMC->OpenFile(url.c_str(), 0);

	if (fileHandle) {
		XBMC->Log(LOG_DEBUG, "Load Complete");
		char buffer[1024];
		while (int bytesRead = XBMC->ReadFile(fileHandle, buffer, 1024))
			retData.append(buffer, bytesRead);
		XBMC->CloseFile(fileHandle);
	}
	return retData;
}

bool PVRLanetTVData::LoadChannelsData(void)
{
	std::string channels_list = LoadFile(g_strListPath);

	Json::Value channels;
	Json::Reader channels_reader;

	bool parsedSuccess = channels_reader.parse(channels_list, channels, false);
	if (!parsedSuccess) {
		XBMC->Log(LOG_ERROR, "Unable to load playlist file '%s':", channels_reader.getFormattedErrorMessages().c_str());
		return false;
	}

	/* Identify Content Pack */
	m_strListPack = channels["pack"].asString();

	/* load channels */
	int channelList_size = channels["list"].size();

	int iUniqueChannelId = 0;
	for (int i = 0; i < channelList_size; ++i)
	{
		CStdString				strTmp;
		PVRLanetTVChannel		channel;

		channel.iUniqueId = ++iUniqueChannelId;
		channel.iEncryptionSystem = 0;

		Json::Value channelData = channels["list"][i];

		/* channel name */
		if (channelData.isMember("title"))
		{
			strTmp = channelData["title"].asString().c_str();
			XBMC->Log(LOG_DEBUG, "Load channel %s", XBMC->UnknownToUTF8(strTmp));
			channel.strChannelName = XBMC->UnknownToUTF8(strTmp);
		}
		else
		{
			continue;
		}

		/* channel number */
		channel.iChannelNumber = atoi(channelData.get("id", "0").asString().c_str());

		/* channel logo */
		channel.strIconPath = g_strLogoPath + channelData.get("id", "0").asString() + ".png";

		/* stream url */
		channel.strStreamURL = channels["edge"].asString() + channelData.get("id", "0").asString() + ".m3u8";

		m_channels.push_back(channel);
	}

	return true;
}

int PVRLanetTVData::GetChannelsAmount(void)
{
  return m_channels.size();
}

PVR_ERROR PVRLanetTVData::GetChannels(ADDON_HANDLE handle, bool bRadio)
{
	for (unsigned int iChannelPtr = 0; iChannelPtr < m_channels.size(); iChannelPtr++)
	{
		PVRLanetTVChannel &channel = m_channels.at(iChannelPtr);

		PVR_CHANNEL xbmcChannel;
		memset(&xbmcChannel, 0, sizeof(PVR_CHANNEL));

		xbmcChannel.iUniqueId					= channel.iUniqueId;
		xbmcChannel.bIsRadio					= false;
		xbmcChannel.iChannelNumber				= channel.iChannelNumber;
		xbmcChannel.iSubChannelNumber			= channel.iChannelNumber;
		xbmcChannel.iEncryptionSystem			= channel.iEncryptionSystem;
		xbmcChannel.bIsHidden					= false;

		strncpy(xbmcChannel.strChannelName, channel.strChannelName.c_str(), sizeof(xbmcChannel.strChannelName) - 1);
		strncpy(xbmcChannel.strStreamURL, channel.strStreamURL.c_str(), sizeof(xbmcChannel.strStreamURL) - 1);
		strncpy(xbmcChannel.strIconPath, channel.strIconPath.c_str(), sizeof(xbmcChannel.strIconPath) - 1);

		PVR->TransferChannelEntry(handle, &xbmcChannel);
}

	return PVR_ERROR_NO_ERROR;
}

bool PVRLanetTVData::GetChannel(const PVR_CHANNEL &channel, PVRLanetTVChannel &myChannel)
{
  for (unsigned int iChannelPtr = 0; iChannelPtr < m_channels.size(); iChannelPtr++)
  {
    PVRLanetTVChannel &thisChannel = m_channels.at(iChannelPtr);
    if (thisChannel.iUniqueId == (int) channel.iUniqueId)
    {
      myChannel.iUniqueId         = thisChannel.iUniqueId;
      myChannel.iChannelNumber    = thisChannel.iChannelNumber;
      myChannel.iSubChannelNumber = thisChannel.iSubChannelNumber;
      myChannel.iEncryptionSystem = thisChannel.iEncryptionSystem;
      myChannel.strChannelName    = thisChannel.strChannelName;
      myChannel.strIconPath       = thisChannel.strIconPath;
      myChannel.strStreamURL      = thisChannel.strStreamURL;

      return true;
    }
  }

  return false;
}


PVR_ERROR PVRLanetTVData::GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd)
{
	if (m_iEpgStart == -1)
		m_iEpgStart = iStart;

	time_t iLastEndTime = m_iEpgStart + 1;
	int iAddBroadcastId = 0;
	XBMC->Log(LOG_DEBUG, "GetEPGForChannel start '%d' end '%d'", iStart, iEnd);

	for (unsigned int iChannelPtr = 0; iChannelPtr < m_channels.size(); iChannelPtr++)
	{
		PVRLanetTVChannel &myChannel = m_channels.at(iChannelPtr);
		if (myChannel.iUniqueId != (int) channel.iUniqueId)
			continue;

		XBMC->Log(LOG_DEBUG, "Try load EPG for channel: %s", myChannel.strChannelName.c_str());

		time_t iLastEndTimeTmp = 0;

		std::string strChannelEPG = LoadFile(g_strEPGPath + std::to_string(myChannel.iChannelNumber).c_str() + "/kodi");

		Json::Value epg;
		Json::Reader epg_reader;

		bool parsedSuccess = epg_reader.parse(strChannelEPG, epg, false);
		if (!parsedSuccess) {
			XBMC->Log(LOG_ERROR, "Unable to load epg file '%s':", epg_reader.getFormattedErrorMessages().c_str());
			continue;
		}
		int iBroadCastId = 0;

		for (unsigned int iEntryPtr = 0; iEntryPtr < epg.size(); iEntryPtr++)
		{

			PVRLanetTVEpgEntry entry;
			Json::Value epgData = epg[iEntryPtr];

			/* broadcast id */
			entry.iBroadcastId = ++iBroadCastId;

			/* channel id */
			entry.iChannelId = myChannel.iUniqueId;

			/* title */
			entry.strTitle = XBMC->UnknownToUTF8(epgData.get("title", "None").asString().c_str());

			/* start */
			entry.startTime = atoi(epgData.get("start", "0").asString().c_str());

			/* end */
			entry.endTime = atoi(epgData.get("stop", "0").asString().c_str());

			/* plot(description) */
			entry.strPlot = XBMC->UnknownToUTF8(epgData.get("text", "None").asString().c_str());

			/* genre */
			std::string genre = epgData.get("cat", "0:0").asString().c_str();
			entry.iGenreType = atoi(genre.substr(0, genre.find(":")).c_str())*16;
			entry.iGenreSubType = atoi(genre.substr(genre.find(":"), genre.length()).c_str());

			/* entry.strGenreDescription = XBMC->UnknownToUTF8("Unknown"); */

			/* likes */
			entry.iStarRating = atoi(epgData.get("likes", "0").asString().c_str());

			XBMC->Log(LOG_DEBUG, "loaded EPG entry '%s' channel '%d' start '%d' end '%d'", entry.strTitle.c_str(), entry.iChannelId, entry.startTime, entry.endTime);

			EPG_TAG tag;
			memset(&tag, 0, sizeof(EPG_TAG));

			tag.iUniqueBroadcastId = entry.iBroadcastId + iAddBroadcastId;
			tag.strTitle = XBMC->UnknownToUTF8(entry.strTitle.c_str());
			tag.iChannelNumber = entry.iChannelId;
			tag.startTime = entry.startTime;
			tag.endTime = entry.endTime;
			tag.strPlot = entry.strPlot.c_str();
			tag.strIconPath = myChannel.strIconPath.c_str();
			tag.iGenreType = entry.iGenreType;
			tag.iGenreSubType = entry.iGenreSubType;
			//tag.strGenreDescription = entry.strGenreDescription.c_str();
			tag.iStarRating = entry.iStarRating;

			iLastEndTimeTmp = tag.endTime;

			PVR->TransferEpgEntry(handle, &tag);
		}

		iLastEndTime = iLastEndTimeTmp;
		XBMC->Log(LOG_DEBUG, "loaded EPG all entry %d", epg.size());
		iAddBroadcastId += epg.size();
	}
	return PVR_ERROR_NO_ERROR;
}
